Index: util/grub-mkconfig_lib.in
===================================================================
--- util/grub-mkconfig_lib.in	(revision 2482)
+++ util/grub-mkconfig_lib.in	(working copy)
@@ -185,8 +185,8 @@ version_test_numeric ()
 
 version_test_gt ()
 {
-  local a=`echo $1 | sed -e "s/[^-]*-//"`
-  local b=`echo $2 | sed -e "s/[^-]*-//"`
+  local a=`echo $1 | sed -e "s/[^-]*-//;s/[._-]\(pre\|rc\|test\|git\|old\)/~\1/g"`
+  local b=`echo $2 | sed -e "s/[^-]*-//;s/[._-]\(pre\|rc\|test\|git\|old\)/~\1/g"`
   local cmp=gt
   if [ "x$b" = "x" ] ; then
     return 0
@@ -196,7 +196,7 @@ version_test_gt ()
     *.old:*) a=`echo -n $a | sed -e s/\.old$//` ; cmp=gt ;;
     *:*.old) b=`echo -n $b | sed -e s/\.old$//` ; cmp=ge ;;
   esac
-  version_test_numeric $a $cmp $b
+  dpkg --compare-versions "$a" $cmp "$b"
   return $?
 }
 
